src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,188 @@
+        for (T toAdd : list) {
+            add(toAdd);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getAt(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        final Node<T> found = getAt(index);
+        final T oldValue = found.getValue();
+        found.setValue(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> found = getAt(index);
+        unlinc(found);
+        return found.getValue();
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            T value = current.getValue();
+            if (value == null && object == null
+                    || value != null && value.equals(object)) {
+                unlinc(current);
+                return true;
+            }
+            current = current.getNext();
+        }
         return false;
     }
 
+    private void unlinc(Node<T> found) {
+        if (found.getNext() == null && found.getPrevious() == null) {
+            setHead(null);
+            setTail(null);
+        } else if (found.getPrevious() == null) {
+            head = found.getNext();
+            head.setPrevious(null);
+        } else if (found.getNext() == null) {
+            tail = found.getPrevious();
+            tail.setNext(null);
+        } else {
+            found.getPrevious().setNext(found.getNext());
+            found.getNext().setPrevious(found.getPrevious());
+        }
+        size--;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getAt(int index) {
+        checkBounds(index);
+        Node<T> result;
+        int middle = (size - 1) / 2;
+        if (index < middle) {
+            result = head;
+            for (int i = 0; i < index; i++) {
+                result = result.getNext();
+            }
+        } else {
+            result = tail;
+            for (int i = 0; i < size - (index + 1); i++) {
+                result = result.getPrevious();
+            }
+        }
+        return result;
+    }
+
+    private void checkBounds(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(
+                    String.format("Index %d out of range 0 - %d", index, size)
+            );
+        }
+    }
+
+    private void checkHeadTail(Node<T> node) {
+        if (node.getNext() == null) {
+            setTail(node);
+        }
+        if (node.getPrevious() == null) {
+            setHead(node);
+        }
+    }
+
+    public void setHead(Node<T> head) {
+        this.head = head;
+    }
+
+    public void setTail(Node<T> tail) {
+        this.tail = tail;
+    }
```

Comment:
I don't think you need these methods, because they perform a rather primitive function and make the code a bit confusing. But I leave the decision up to you. But they must be private because the user can add any head or tail and the links will not be redefined